var/spack/repos/builtin/packages/adios2/package.py

```diff
@@ -281,3 +284,50 @@ def setup_run_environment(self, env):
             env.prepend_path("HDF5_PLUGIN_PATH", os.path.dirname(all_libs[idx]))
         except ValueError:
             pass
+
+    @run_after("install")
+    def setup_install_tests(self):
+        """
+        Copy the example files after the package is installed to an
+        install test subdirectory for use during `spack test run`.
+        """
+        extra_install_tests = [join_path("testing", "install", "C")]
+        self.cache_extra_test_sources(extra_install_tests)
+
+    def test_run_executables(self):
+        """Run installed adios2 executables"""
+
+        commands_and_args = [("bpls", ["-v", "-V"])]
+
+        for cmd, opts in commands_and_args:
+            with test_part(
+                self, f"test_adios2_cmds_{cmd}", purpose=f"run installed adios2 executable {cmd}"
+            ):
+                exe = which(join_path(self.prefix.bin, cmd))
+                exe(*opts)
+
+    def test_build_example(self):
+        """Build and run an example program"""
+        src_dir = self.test_suite.current_test_cache_dir.testing.install.C
+        test_stage_dir = self.test_suite.test_dir_for_spec(self.spec)
+
+        # Create the build tree within this spec's test stage dir so it gets
+        # cleaned up automatically
+        build_dir = tempfile.mkdtemp(dir=test_stage_dir)
+
+        std_cmake_args = []
+
+        if "mpi" in self.spec:
+            mpi_exec = join_path(self.spec["mpi"].prefix, "bin", "mpiexec")
+            std_cmake_args.append("-DMPIEXEC_EXECUTABLE={0}".format(mpi_exec))
```

Comment:
@tldahlgren In my local testing, I needed to add this hint to help cmake find the spack-built mpi, otherwise compilation failed due to finding an mpi on my local system.

So that's making me wonder if there's something more general I need to do, so that only spack-built things are used to build the tests?

Reply:
Good question that I can relate to as I'm on my fourth attempt to build the package so I can run the tests to check o a couple of things but am (still) using externals to avoid having to build everything from scratch.  Will likely abandon that on the next round.

In general tests should use the same tools that were used during the build.  So maybe try `spack install -U`  or `spack install --fresh`?

Reply:
I was able to build the software with a system (so externally installed) `gcc@11.2.1` and the `cmake` used by the test was the same used by the build.  

It's not obvious to me why it wasn't necessary to create the `cmake` executable instance in the package though.